### PR TITLE
Add basic support for v2 hashes in transmission-show

### PR DESCRIPTION
**Pull request by Lanzaa**
Relates to issue #458 and [BEP-0052](http://bittorrent.org/beps/bep_0052.html) (bittorrent v2).

Included in this patch is support for calculating a v2 info hash from a torrent file.

I am hoping for comments as I continue to work on further v2 support. I have not done much C++ programming in a while, so suggestions for improvement are very welcome.

### CODE REVIEW

**Comment by Lanzaa** [comment]
Ok, I've got a basic update with the other crypto libraries and added a test case which will verify the very basics of sha256 support.

Hopefully it all works, I did the change blind w/o a build environment for them. I don't really want to spend much time on this portion.

File: libtransmission/crypto-utils.h [hunk 1]
```diff
@@ -186,10 +226,20 @@ std::string tr_base64_decode(std::string_view input);
 std::string tr_sha1_to_string(tr_sha1_digest_t const&);
 
 /**
- * @brief Generate a sha1 digest from a hex string.
+ * @brief Generate a sha256 digest from a hex string.
```
Comment (nevack) [inline libtransmission/crypto-utils.h:229]: typo?

File: libtransmission/crypto-utils.h [hunk 2]
```diff
@@ -70,6 +72,44 @@ std::optional<tr_sha1_digest_t> tr_sha1(T... args)
     return std::nullopt;
 }
 
+/**
+ * @brief Allocate and initialize new SHA1 hasher context.
```
Comment (nevack) [inline libtransmission/crypto-utils.h:76]:
```suggestion
 * @brief Allocate and initialize new SHA1 hasher context.
 * @brief Allocate and initialize new SHA256 hasher context.
```

File: libtransmission/crypto-utils.h [hunk 3]
```diff
@@ -70,6 +72,44 @@ std::optional<tr_sha1_digest_t> tr_sha1(T... args)
...
+ */
+tr_sha256_ctx_t tr_sha256_init(void);
+
+/**
+ * @brief Update SHA1 hash.
```
Comment (nevack) [inline libtransmission/crypto-utils.h:81]:
```suggestion
 * @brief Update SHA256 hash.
```

File: libtransmission/crypto-utils.h [hunk 4]
```diff
@@ -70,6 +72,44 @@ std::optional<tr_sha1_digest_t> tr_sha1(T... args)
...
+ */
+bool tr_sha256_update(tr_sha256_ctx_t handle, void const* data, size_t data_length);
+
+/**
+ * @brief Finalize and export SHA1 hash, free hasher context.
```
Comment (nevack) [inline libtransmission/crypto-utils.h:86]:
```suggestion
 * @brief Finalize and export SHA256 hash, free hasher context.
```

File: libtransmission/crypto-utils.h (hunk 1 above)
Comment (nevack) [reply libtransmission/crypto-utils.h:229]:
```suggestion
 * @brief Generate a sha1 digest from a hex string.
```

File: libtransmission/makemeta.h [hunk 5]
```diff
@@ -6,6 +6,7 @@
 #pragma once
 
 #include <cstdint> // uint32_t, uint64_t
+#include <optional>
```
Comment (nevack) [inline libtransmission/makemeta.h:9]: Is this include needed here?

File: libtransmission/crypto-utils.h [hunk 6]
```diff
@@ -70,6 +72,44 @@ std::optional<tr_sha1_digest_t> tr_sha1(T... args)
...
+ */
+std::optional<tr_sha256_digest_t> tr_sha256_final(tr_sha256_ctx_t handle);
+
+/**
+ * @brief generate a SHA256 hash from some memory
+ */
+template<typename... T>
+std::optional<tr_sha256_digest_t> tr_sha256(T... args)
```
Comment (nevack) [inline libtransmission/crypto-utils.h:94]:
```
31][Step 2/4] In file included from /usr/home/user/buildAgent4/work/1f3bff43dd55cfe9/tests/libtransmission/crypto-test.cc:19:
[21:48:31][Step 2/4] In file included from /usr/home/user/buildAgent4/work/1f3bff43dd55cfe9/tests/libtransmission/crypto-test-ref.h:65:
[21:48:31][Step 2/4] In file included from /usr/home/user/buildAgent4/work/1f3bff43dd55cfe9/libtransmission/crypto.h:18:
[21:48:31][Step 2/4] /usr/home/user/buildAgent4/work/1f3bff43dd55cfe9/libtransmission/crypto-utils.h:94:35: error: redefinition of 'tr_sha256'
[21:48:31][Step 2/4] std::optional<tr_sha256_digest_t> tr_sha256(T... args)
[21:48:31][Step 2/4]                                   ^
[21:48:31][Step 2/4] /usr/home/user/buildAgent4/work/1f3bff43dd55cfe9/libtransmission/crypto-utils.h:94:35: note: previous definition is here
[21:48:31][Step 2/4] std::optional<tr_sha256_digest_t> tr_sha256(T... args)
[21:48:31][Step 2/4]                         
```

File: libtransmission/crypto-utils-polarssl.cc [hunk 7]
```diff
@@ -178,6 +180,54 @@ std::optional<tr_sha1_digest_t> tr_sha1_final(tr_sha1_ctx_t raw_handle)
 ****
 ***/
 
+tr_sha256_ctx_t tr_sha256_init(void)
+{
+    api_sha256_context* handle = tr_new0(api_sha256_context, 1);
+
+#if API_VERSION_NUMBER >= 0x01030800
```
Comment (nevack) [inline libtransmission/crypto-utils-polarssl.cc:187]: nit: It's better to leave comment here and there what version this constant maps to.

File: libtransmission/magnet-metainfo.cc [hunk 8]
```diff
@@ -151,6 +151,19 @@ std::optional<tr_sha1_digest_t> parseHash(std::string_view sv)
...
+{
+    // http://bittorrent.org/beps/bep_0009.html
+    // Is the info-hash v2 hex encoded and tag removed, for a total of 64 characters.
+
+    if (auto const hash = tr_sha256_from_string(sv); hash)
+    {
+        return hash;
+    }
+
+    return {};
```
[the remark this answers is not in the document]
Comment (ckerr) [reply libtransmission/magnet-metainfo.cc:164]: FWIW I'm pretty sure I've been inconsistent in my usage too. It's possible that this code is copied from inconsistent use in the sha1() function :smile:

File: libtransmission/crypto-utils.h (hunk 6 above)
Comment (ckerr) [reply libtransmission/crypto-utils.h:94]: This is probably due to misunderstanding of how the crypto- files work. The design predates the C++ migration and has an odd way of plugging in implementations via `#include`. Totally out-of-scope for this PR but if anyone wanted to make a pure virtual interface class for the crypto API and then have impl .cc filefs for the different providers that would be :100:

**Comment by ckerr** [comment]
> I am hoping for comments as I continue to work on further v2 support. 

If you have any questions or need anything, feel free to open new issues / discussions and assign me or @ me. I love that you're doing this.

File: libtransmission/crypto-utils.h (hunk 6 above)
Comment (Lanzaa) [reply libtransmission/crypto-utils.h:94]: I don't know how to fix that. @ckerr help?

File: libtransmission/crypto-utils.h (hunk 1 above)
Comment (Lanzaa) [reply libtransmission/crypto-utils.h:229]: Copy/paste error.

File: libtransmission/magnet-metainfo.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (Lanzaa) [reply libtransmission/magnet-metainfo.cc:164]: Yeah, returning '{}' is consistent with the rest of the file. I'm not sure which is preferred.

File: libtransmission/crypto-utils.h (hunk 4 above)
Comment (Lanzaa) [reply libtransmission/crypto-utils.h]: Copy/paste error.

File: libtransmission/crypto-utils.h (hunk 3 above)
Comment (Lanzaa) [reply libtransmission/crypto-utils.h]: Copy/paste error.

File: libtransmission/crypto-utils.h (hunk 2 above)
Comment (Lanzaa) [reply libtransmission/crypto-utils.h]: Copy/paste error.

File: libtransmission/makemeta.h (hunk 5 above)
Comment (Lanzaa) [reply libtransmission/makemeta.h]: Nope.

File: libtransmission/crypto-utils-polarssl.cc (hunk 7 above)
Comment (Lanzaa) [reply libtransmission/crypto-utils-polarssl.cc:187]: Eh, I copy pasted this from the sha1 stuff. My vote would be to bump the minimum polarssl/mbedtls version and remove the api version check from both places. I bet the constant maps simply to [1.3.8 (changelog)](https://github.com/Mbed-TLS/mbedtls/blob/1dc6848679982dde8b7cd42a547358eb04fed6ca/ChangeLog#L3516) . Version 1.3.8 was released in 2014.

I honestly don't want to do that work.

File: libtransmission/magnet-metainfo.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (ckerr) [reply libtransmission/magnet-metainfo.cc:164]: Generally I use `{}`. There are a couple of places in the codebase that use `std::nullopt`, e.g. when returning a ternary where `{}` is not allowed in a ternary

**Comment by GaryElshaw** [comment]
That change-log might need a slight revision, huh ;-)

**Comment by ckerr** [comment]
> That change-log might need a slight revision, huh ;-)

The nice thing about `--fixup and --autosquash` is that it boils down into a single commit when you rebase.

Since I almost always squash & merge to `main` it all comes in as a single commit anyway :smile_cat:

**Comment by ckerr** [comment]
Looks like 23e2390 might've finally gotten it right, but one of the CI boxes is offline so this will probably need to wait until sometime tomorrow when we can get CI green again. :neutral_face:

**Comment by ckerr** [comment]
OK, that one runner's been down for a day and a half now and the other runs were all green. I think it's OK to go ahead with this.

**Pull request merged**